### ISSUE TITLE
Trimming Support

### ISSUE THREAD

**Issue by CasualPokePlayer** [TASEmulators/BizHawk#3767]
Something to look at wrt .NET6 (see #2261). One benefit of .NET6 is the ability to do a "self-contained" build which packs in all the framework and runtime (so no need for a prereq installer) (see also #1415). This however makes the build much larger. One way to reduce this size is to [enable trimming](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-self-contained), which removes unused code. (perhaps this might even give us an overall size reduction compared to a framework deployed build?)

However, trimming comes with [several limitations](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/incompatibilities):
- Built-in COM marshalling
	- Only used directly in BizHawk.Common, in [ShellLinkImports.cs](https://github.com/TASEmulators/BizHawk/blob/a351dcc55795ed6df8f62f6fb16f4be031fb27c3/src/BizHawk.Common/Win32/ShellLinkImports.cs), [Win32Imports.cs](https://github.com/TASEmulators/BizHawk/blob/a351dcc55795ed6df8f62f6fb16f4be031fb27c3/src/BizHawk.Common/Win32/Win32Imports.cs), and [Win32ShellContextMenu.cs](https://github.com/TASEmulators/BizHawk/blob/a351dcc55795ed6df8f62f6fb16f4be031fb27c3/src/BizHawk.Common/Win32/Win32ShellContextMenu.cs).
	- Maybe used in some deps wrapping around DirectX APIs? (DirectSound/XAudio2/DirectInput/XInput/Direct3D9?)
	- WinForms uses COM marshalling heavily so WinForms projects can't be trimmed (BizHawk.Client.EmuHawk and BizHawk.Bizware.Graphics.Controls). Another reason to move away from WinForms.
- WPF
	-  Not used by BizHawk, but something to keep in mind when looking for WinForms alternatives, as some might just wrap around WPF on Windows (e.g. [Eto](https://github.com/picoe/Eto)), and so can't be trimmed.
	- Similarly we might not want to use UI frameworks which wrap around WinForms on Windows for the same reason.
- Reflection
	- Reflection itself can often be made [compatible with trimming](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/fixing-warnings), although we probably want to move away from reflection wherever possible.
	- Serializers (i.e. Newtonsoft.Json) often can't be made compatible however, but it should be easy enough to move to source generator alternatives (i.e. [System.Text.Json](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation)).
- Dynamic assembly loading and execution
	- i.e. External Tools, so we can't have External Tools with trimming?
	- Perhaps we could rework ApiHawk so it delivers APIs with any native library with a pre-defined C interface which allows them to receive APIs, making external tools language agnostic too (something [native exports](https://devblogs.microsoft.com/dotnet/announcing-net-5-0/#native-exports) for .NET tools?)
	- Perhaps there's some workaround here? Can External Tools just package in the .NET framework/runtime code that they need? Would that create some conflict with BizHawk's own self contained framework/runtime code?

**Comment by YoshiRulz** [comment]
Worth noting that trimming is configurable; you can [explicitly include symbols](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0#root-descriptors) you know will be used.

I have [some notes re: reflection](https://gitlab.com/YoshiRulz/yoshis-hawk-thoughts/-/issues/15) and [publishing/packaging](https://gitlab.com/YoshiRulz/yoshis-hawk-thoughts/-/issues/101).

**Comment by Morilli** [comment]
>  WinForms projects can't be trimmed (BizHawk.Client.EmuHawk [...]

So this means it cannot be enabled AT ALL until we move away from winforms, right?

**Comment by CasualPokePlayer** [comment]
I assume trimming can be controlled on a per project basis here? I assume regardless however we can't go enable AOT until the entirety of BizHawk is trimmable.

**Comment by Morilli** [comment]
Yeah I don't think trimming is gonna happen any time soon. Reflection is just one example of many patterns that are not compatible with trimming and are extremely hard to fix at the very least. If you haven't you could try to enable trimming on the .NET6 targeting branch and see how many errors you get.

**Comment by CasualPokePlayer** [comment]
The trimming is somewhat weirdish. For each project you want to manually specify it in, you need to have at least .NET 6 targeted. However, in reality this isn't really needed per se (unless you're doing partial trimming and probably for warnings), since if you perform *full trimming*, then specifying that for the exe project is enough for all assemblies used to be trimmed (even .NET Standard/.NET 5 and below ones).
